sampler.go

```diff
@@ -184,13 +184,125 @@ func (s *rateLimitingSampler) Equal(other Sampler) bool {
 	return false
 }
 
-// Tags implements Tags of Sampler
-func (s *rateLimitingSampler) getTags() []tag {
+// getTags implements getTags of Sampler
+func (s *rateLimitingSampler) getTags(operation string) []tag {
 	return s.tags
 }
 
 // -----------------------
 
+type adaptiveSampler struct {
+	rateLimitingSamplers  map[string]Sampler
+	probabilisticSamplers map[string]Sampler
+	tags                  map[string][]tag
+
+	// TODO expand to keep track of newly seen operations
+}
+
+type adaptiveSamplingRates struct {
+	samplingRate       *float64
+	maxTracesPerSecond *float64
+}
+
+// NewAdaptiveSampler adaptiveSampler is a delegating sampler that applies both probabilisticSampler and
+// rateLimitingSampler per operation. The probabilisticSampler is given higher priority when tags are emitted,
+// ie. if IsSampled() for both samplers return true, the tags for probabilisticSampler will be used.
+//
+// This sampler assumes that it will be used by RemotelyControlledSampler hence it is up to the user to
+// ensure this is thread-safe if not used by RemotelyControlledSampler.
+func NewAdaptiveSampler(samplingRates map[string]*adaptiveSamplingRates) (Sampler, error) {
+	probabilisticSamplers := make(map[string]Sampler)
+	rateLimitingSamplers := make(map[string]Sampler)
+	for operation, rates := range samplingRates {
+		if rates.samplingRate != nil {
+			if sampler, err := NewProbabilisticSampler(*rates.samplingRate); err != nil {
+				return nil, err
+			} else {
+				probabilisticSamplers[operation] = sampler
+			}
+		}
+		if rates.maxTracesPerSecond != nil {
+			if sampler, err := NewRateLimitingSampler(*rates.maxTracesPerSecond); err != nil {
+				return nil, err
+			} else {
+				rateLimitingSamplers[operation] = sampler
+			}
+		}
+	}
+	tags := make(map[string][]tag, len(samplingRates))
+	for operation := range samplingRates {
+		tags[operation] = []tag{}
+	}
+	return &adaptiveSampler{
+		probabilisticSamplers: probabilisticSamplers,
+		rateLimitingSamplers:  rateLimitingSamplers,
+		tags:                  tags,
+	}, nil
+}
+
+func (s *adaptiveSampler) IsSampled(id uint64, operation string) bool {
+	probabilisticSampler := s.probabilisticSamplers[operation]
+	if probabilisticSampler != nil && probabilisticSampler.IsSampled(id, operation) {
+		s.tags[operation] = probabilisticSampler.getTags(operation)
```

Comment:
if you change the signature to return tags along with the decision, then the adaptive sampler does not need its own tags, it simply returns the tags from the sampler that was used